aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,680 @@
+  /*
+    See ATen/native/sparse/cpu/Softmax.cpp:cpu_sparse_coo_softmax_backward for
+    the CPU implementation of the sparse softmax backward algorithm that this
+    implementation is based on.
+  */
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int index = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  while (index < size) {
+    int64_t offset = pool_offsets[index];
+    int64_t* pool_indices = sorted_pool_indices + offset;
+    int64_t pool_indices_size = pool_sizes[index];
+
+    scalar_t* tmp_row = exp_sums_rows + index * nvalues;
+
+    /* Compute tmp = - sum_j output_j * grad_j */
+    for (int64_t p = 0; p < pool_indices_size; p++) {
+      auto i = pool_indices[p];
+      auto out_values_row = out_values_accessor[i];
+      auto values_row = values_accessor[i];
+      auto j = lower_bound_values[i];
+      if (j < grad_nnz && (out_offsets[i] == grad_offsets[j])) {
+        auto grad_values_row = grad_values_accessor[j];
+        for (int64_t k = 0; k < nvalues; k++) {
+          if (LogSoftMax) {
+            tmp_row[k] -= grad_values_row[k];
+          } else {
+            tmp_row[k] -= out_values_row[k] * grad_values_row[k];
+          }
+        }
+      }
+    }
+    /* Compute grad_input = output * (grad + tmp)*/
+    for (int64_t p = 0; p < pool_indices_size; p++) {
+      auto i = pool_indices[p];
+      auto out_values_row = out_values_accessor[i];
+      auto values_row = values_accessor[i];
+      auto j = lower_bound_values[i];
+
+      if (j < grad_nnz && (out_offsets[i] == grad_offsets[j])) {
+        auto grad_values_row = grad_values_accessor[j];
+        for (int64_t k = 0; k < nvalues; k++) {
+          if (LogSoftMax) {
+            values_row[k] = grad_values_row[k] +
+                c10::cuda::compat::exp(out_values_row[k]) * tmp_row[k];
+          } else {
+            values_row[k] =
+                out_values_row[k] * (grad_values_row[k] + tmp_row[k]);
+          }
+        }
+      } else {
+        for (int64_t k = 0; k < nvalues; k++) {
+          if (LogSoftMax) {
+            values_row[k] =
+                c10::cuda::compat::exp(out_values_row[k]) * tmp_row[k];
+          } else {
+            values_row[k] = out_values_row[k] * tmp_row[k];
+          }
+        }
+      }
+    }
+    index += step;
+  }
+}
+
+using thrust_ptr = thrust::device_ptr<int64_t>;
+
+Tensor get_offsets(
+    const Tensor& indices,
+    const IntArrayRef& sizes,
+    const int64_t dim) {
+  /*
+    See ATen/native/sparse/cpu/Softmax.cpp:get_offsets for the CPU
+    implementation of get_offsets function that this implementation is based on.
+  */
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  auto ndim = indices.size(0);
+  auto nnz = indices.size(1);
+  auto host_strides =
```

Comment:
Creating a CPU tensor here seems excessive. What about creating a vector and using something like std::accumulate in place of the loop below?

Reply:
(you probably need to use `from_blob` to construct a tensor from a vector)